src/libs/kata-types/src/config/hypervisor/mod.rs

```diff
@@ -0,0 +1,1061 @@
+    #[serde(default)]
+    pub valid_file_mem_backends: Vec<String>,
+
+    /// Enable pre allocation of VM RAM, default false
+    ///
+    /// Enabling this will result in lower container density as all of the memory will be allocated
+    /// and locked. This is useful when you want to reserve all the memory upfront or in the cases
+    /// where you want memory latencies to be very predictable
+    #[serde(default)]
+    pub enable_mem_prealloc: bool,
+
+    /// Enable huge pages for VM RAM, default false
+    ///
+    /// Enabling this will result in the VM memory being allocated using huge pages. This is useful
+    /// when you want to use vhost-user network stacks within the container. This will automatically
+    /// result in memory pre allocation.
+    #[serde(default)]
+    pub enable_hugepages: bool,
+
+    /// Specifies virtio-mem will be enabled or not.
+    ///
+    /// Please note that this option should be used with the command
+    /// "echo 1 > /proc/sys/vm/overcommit_memory".
+    #[serde(default)]
+    pub enable_virtio_mem: bool,
+
+    /// Enable swap of vm memory. Default false.
+    ///
+    /// The behaviour is undefined if mem_prealloc is also set to true
+    #[serde(default)]
+    pub enable_swap: bool,
+
+    /// Enable swap in the guest. Default false.
+    ///
+    /// When enable_guest_swap is enabled, insert a raw file to the guest as the swap device if the
+    /// swappiness of a container (set by annotation "io.katacontainers.container.resource.swappiness")
+    /// is bigger than 0.
+    ///
+    /// The size of the swap device should be swap_in_bytes (set by annotation
+    /// "io.katacontainers.container.resource.swap_in_bytes") - memory_limit_in_bytes.
+    /// If swap_in_bytes is not set, the size should be memory_limit_in_bytes.
+    /// If swap_in_bytes and memory_limit_in_bytes is not set, the size should be default_memory.
+    #[serde(default)]
+    pub enable_guest_swap: bool,
+}
+
+impl MemoryInfo {
+    /// Adjust the configuration information after loading from configuration file.
+    pub fn adjust_configuration(&mut self) -> Result<()> {
+        resolve_path!(
+            self.file_mem_backend,
+            "Memory backend file {} is invalid: {}"
+        )?;
+        Ok(())
+    }
+
+    /// Validate the configuration information.
+    pub fn validate(&self) -> Result<()> {
+        validate_path!(
+            self.file_mem_backend,
+            "Memory backend file {} is invalid: {}"
+        )?;
+        if self.default_memory == 0 {
+            return Err(eother!("Configured memory size for guest vm is zero"));
+        }
+        if self.memory_slots == 0 {
+            return Err(eother!("Configured memory slots for guest vm is zero"));
+        }
+
+        Ok(())
+    }
+
+    /// Validate path of memory backend files.
+    pub fn validate_memory_backend_path<P: AsRef<Path>>(&self, path: P) -> Result<()> {
+        validate_path_pattern(&self.valid_file_mem_backends, path)
```

Comment:
Shouldn't this just be in `validate`?

Reply:
this function is used to validate annotation info which will be called in update_config_by_annotation() in annotations/mods.rs